### ISSUE TITLE
Simplify and clarify how to open a GraphQL session

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#3830]
## Problem

In order for a client library or tool to communicate with the Dagger Engine, it must open a GraphQL session. When using a Dagger SDK this is abstracted away. However some developers need to target the GraphQL API without the help of an SDK. At the moment this is a painful experience because, although Dagger's GraphQL implementation is standard, the procedure for *opening a session* is specific to Dagger - and it is quite painful to understand and implement.

Specifically, the procedure for opening a graphql session has 3 problems:

1) It's not documented: a client developer must ask questions and read other implementations to learn it
2) It's misleading: learning the current interface creates misconceptions about Dagge's architecture.
3) It's too much work to implement: when multiplied by many clients down the line, even small increments in implementation effort can make a difference.

### People affected

As noted above:

> some developers need to target the GraphQL API without the help of an SDK

This includes:

* Developers using languages for which a Dagger SDK is not yet available
* Developers of unofficial SDKs and exec wrappers
* Contributors to Dagger SDKs
* Contributors to the Dagger CLI, specifically  [the `dagger run` wrapper](https://github.com/dagger/dagger/issues/3347)

Additionally, some people may want to understand the process for opening a GraphQL session, even if they themselves don't need to implement it. This includes:

* Power users who wish to understand how to operate Dagger in depth before using it in production
* People looking to setup Dagger in a client-server mode, and being misled by the `$DAGGER_HOST` environment variable used by the SDKs.

## Solution

Define a new "Dagger Session Interface" (DSI?)  that 1) is well documented, 2) helps understand Dagger's architecture, and 3) is as easy to implement as possible.

## Design proposal

For a client to open a session to a Dagger Engine, it must implement the *Dagger Session Interface* (DSI). There are two levels of implementations for DSI: basic, and advanced.

### DSI-basic: no bootstrap

DSI-basic defines the simplest possible way to open a session to a Dagger engine. It can typically be implemented in 10 lines of code or less. The tradeoff is that the client cannot be executed standalone: the operator must execute it in the context of a `dagger run` command. For example: `dagger run python mytool.py`, or `dagger go run mypipeline.go`.

To implement DSI-basic, follow the steps below:

```mermaid
graph TD;

subgraph DSI-basic
start[Start]
if_env_set(Are $DAGGER_SESSION_PORT and $DAGGER_SESSION_TOKEN set?)
env_set_connect[connect to tcp://localhost:$DAGGER_SESSION_PORT, as user $DAGGER_SESSION_TOKEN]
error[Error]
success[Success]
end

start --> if_env_set
if_env_set --yes--> env_set_connect --> success
if_env_set --no--> error
```

### DSI-advanced: with bootstrap

DSI-advanced extends DSI-basic with a bootstrap feature: if a session is not already open, then DSI-advanced specifies how to automatically open one, with no additional dependencies. A client implementing DSI-advanced can be executed standalone: it does not need to be wrapped with `dagger run`, and the Dagger CLI does not need to be installed. The tradeoff is that DSI-advanced is more complex to implement.

```mermaid
graph TD;

subgraph DSI-advanced

subgraph DSI-basic
start[Start]
if_env_set(Are $DAGGER_SESSION_PORT and $DAGGER_SESSION_TOKEN set?)
env_set_connect[connect to tcp://localhost:$DAGGER_SESSION_PORT, as user $DAGGER_SESSION_TOKEN]
success[Success]
end

start --> if_env_set
if_env_set --yes--> env_set_connect --> success

if_cli_installed(Is Dagger CLI installed in $PATH?)
cli_auto_install[Download the latest dagger CLI]
cli_session[Execute 'dagger session']
cli_read_env[Read json-encoded environment object from standard output of 'dagger session', export to environment variables]

if_env_set --no--> if_cli_installed
if_cli_installed --yes--> cli_session --> cli_read_env --> env_set_connect
if_cli_installed --no--> cli_auto_install --> cli_session

end
```

### DAGGER_SESSION_PORT or DAGGER_SESSION_URL?

Since the session must always be accessible as a tcp port *on localhost only*, there are two options for the client to determine the URL:

| | Option 1: _PORT | Option 2: _URL |
| -- | -- | -- |
| How to assemble session URL | `http://localhost:$DAGGER_SESSION_PORT` | `$DAGGER_SESSION_URL` |
| In current proposal? | Yes | No |
| Pros | Makes it more clear that the session is local only; no confusion with `$DAGGER_HOST`; less opportunity for ill-advised hacks that will only end in tears | Less characters to type for client developers; more flexibility for unknown future use cases |
| Cons | More characters to type for client developers; less flexibility for unknown future use cases | Less clear that the session is local only; confusion with `$DAGGER_HOST`; more opportunity for ill-advised hacks that will only end in tears |

### Which CLI to download

In DSI-advanced, the client must download the Dagger CLI. But which version?

* Currently, clients in official SDKs specify an exact version
* How does an official client developer decide when to upgrade the hardcoded version, and to what value?
* What about everyone else? Should they also specify a version? If so, which version and how/when to change it? If not: then what?

### Could `dagger listen` be folded into `dagger session`?

Currently these two commands are different (but hidden). They are close enough in functionality, though, that the possibility of merging them is intriguing.

## Implementation Progress
- [x] Update CLI to use engine.Start: https://github.com/dagger/dagger/pull/4119
- [x] Add hidden `session` subcommand to CLI that is exact replica of engine-session bin (to start): https://github.com/dagger/dagger/pull/4145
- [x] Publish CLI on every push to main (same as engine image): https://github.com/dagger/dagger/pull/4208
   * This is needed at minimum to give the SDKs something to pull before we do an official release of the CLI w/ `session` subcommand, but it's probably desired anyways
- [x] Update SDKs to pull CLI from S3 and invoke `session` subcommand: https://github.com/dagger/dagger/pull/4194
   * Also requires bumping the CLI url instead of engine session URL when we publish the engine+cli
- [x] Make SDKs use session tokens (same as `dagger run`): https://github.com/dagger/dagger/pull/4145
   * This will make the SDKs compatible w/ `dagger run`
- [x] Switch `DAGGER_SESSION_URL` -> `DAGGER_SESSION_PORT`: https://github.com/dagger/dagger/pull/4285
- [x] Re-activate macos provisioning tests: https://github.com/dagger/dagger/pull/4208
- [x] Zipfile CLI download support on windows: https://github.com/dagger/dagger/pull/4286
- [x] Change buildkitd state dir to something with control and configure: https://github.com/dagger/dagger/pull/4287
- [x] Add the rest of the builtin runner connection providers we get from buildkit and remove external access to `docker-image` https://github.com/dagger/dagger/pull/4288
- [x] Add docs for all this. Internal for `_EXPERIMENTAL_*` params, external for the others (probably just need updates to [this doc)](https://docs.dagger.io/api/254103/build-custom-client): https://github.com/dagger/dagger/issues/4215

Spun off longer-term follow ups:
- https://github.com/dagger/dagger/issues/4319

**Comment by shykes** [comment]
cc @aluzzardi @sipsma @marcosnils

**Comment by sipsma** [comment]
> In DSI-advanced, the client must download the Dagger Engine OCI image. But which version of the image?

I think this should be viewed as the exact same thing as bumping a dependency (i.e. in go.mod, package.json ,etc.). You should update this when there is an update to the engine that you want to pick up and should obviously test against that new engine before releasing a new version of the SDK.

The engine images specify a digest right now because trying to deal with mutable tags introduces a lot more conditions to handle. However, we still use tags set to the engine version, so it's still easy to find the digest for a given engine version.

The biggest headache around this is the garbage collection of engines w/ different versions honestly. I just opened an issue for improving that in the short-to-medium term: https://github.com/dagger/dagger/issues/3849

Of course, everything I'm writing here should be turned into developer documentation for SDK developers once we have consensus on it all :-)

**Comment by sipsma** [comment]
> @sipsma one small clarifying question before I address the rest of your message: in your proposed flow, the step "Execute dagger-engine-session with --addr $DAGGER_SESSION_ADDR" comes after checking that $DAGGER_SESSION_ADDR is not set. Is that a typo or am I missing something?

Doh... no that was just a mistake, I edited the diagram now with the original intention.

**Comment by sipsma** [comment]
> What do you mean in this context by "override the runner"?

I just mean specify the runner to use. "Override" in this context just meant "use this runner instead of any default". I realize that we currently *don't* have a default in the engine-session binary, I'm just thinking about the future where maybe we do end up with one there.

**Comment by shykes** [comment]
> For the basic case, I think we should not overly tie ourselves to tcp+localhost (I realize the current implementation does that, I'm just thinking longer term now). We've talked about this in the past but to recap for others, sending secrets in plaintext over localhost is not ideal, in the long run we should probably have the option of using unix sockets.
> 
> This doesn't mean every SDK _has_ to implement support for unix, just that it should be an option. So, _if an SDK chooses to also support unix_ this would be the full decision dag:

This is not ideal for a few reasons:

1) It's more implementation work for every client developer. Even if they don't implement both unix and tcp, they have to implement the extra logic to support choosing
2) It's also more cognitive work, because each client has to *choose* between unix and tcp, which means they need to understand the different, study the tradeoffs etc. It would be much easier for everyone if we just choose one way to do it, and standardize on that

In that spirit, a few questions:

* You mention the security issue with sending plaintext secrets over tcp. Could you clarify what exactly the issue is?
* Presumably using unix sockets has a downside, or we would just be using that everywhere. Could you also clarify what the downside is?

Ideally we would look at the tradeoffs, and choose one method now. Then, if in the future we want to extend the spec, we can always add another environment variable.

**Comment by sipsma** [comment]
> You mention the security issue with sending plaintext secrets over tcp

Basically if any one has capabilities to sniff localhost traffic (on Linux I think it's `cap_net_admin`, maybe `cap_net_raw` too) then anyone can read secrets we are passing over it. The fact that this requires capabilities is barely okay as far as safeguards go (which is why I don't think it's an emergency), but in general that isn't really an ideal compromise since there's no reason having those capabilities specifically should break our security model.

We could of course encrypt our traffic over localhost, but that's going to be it's own burden on sdk developers.

Thinking out loud now: I suppose we could also tackle secret encryption separately (i.e. encrypt them before sending them over any wire, localhost or otherwise). And maybe it's inevitable that we need to do that since we don't want to send them in plaintext to a remote runner? Not sure yet, but if we solve encrypting secrets on the wire separately, then my concerns about localhost would be gone.

> Presumably using unix sockets has a downside, or we would just be using that everywhere. Could you also clarify what the downside is?

1. It apparently isn't always straightforward to configure a graphql client to use a unix socket in every language. In Go it's straightforward but I've been told it's less straightforward in python+nodejs. In general, I'm not sure how much of a burden this will be for developers of SDKs in arbitrary languages
2. Unix sockets are supported on linux, macos and even windows, but windows support is somewhat recent and I'm not sure how many languages support it yet. So trying to deal with platform specific issues is probably more painful than with the localhost approach, which is essentially platform agnostic.

Those are all conquerable problems, but the question is if it's worth it to solve them vs accepting the possible security compromises of just always using localhost.

**Comment by aluzzardi** [comment]
Isn't that what `dagger listen` provides, if an external tool wants to probe the GraphQL API directly?

**Comment by shykes** [comment]
> Isn't that what `dagger listen` provides, if an external tool wants to probe the GraphQL API directly?
> 
> If it's for connecting ecosystem tools (graphiql, schema linters, ...) those things are http only. All client libraries we use in each language expects an http:// endpoint.
> 
> So the easy way to connect to the API is either to `dagger listen` on the side and connect there, `dagger listen` from code, or wrapping the command in a `dagger exec`

@aluzzardi is this in response to a particular message in the thread? Or just to the entire proposal? (ie. “why do we need any of this?”)

**Comment by marcosnils** [comment]
catching up with this. As we spoke yesterday and after finishing the first implementation of `dagger exec` [here](https://github.com/dagger/dagger/pull/3748/), I feel good with releasing your proposed workflow as an initial DSI.

Regarding @sipsma comments about security, I agree they're a concern but I don't see why we couldn't release a future implementation with `DAGGER_SESSION_ADDR` when users start requesting for that. 



> > You mention the security issue with sending plaintext secrets over tcp
> 
> Basically if any one has capabilities to sniff localhost traffic (on Linux I think it's `cap_net_admin`, maybe `cap_net_raw` too) then anyone can read secrets we are passing over it. The fact that this requires capabilities is barely okay as far as safeguards go (which is why I don't think it's an emergency), but in general that isn't really an ideal compromise since there's no reason having those capabilities specifically should break our security model.

+1 on Erik's answer here. Since this is advanced mode, I'd expect these users to be more aware of they image that they want/need to pull. Not sure if there's anything we should do here besides clearly documenting API and engine releases.

> Currently the official clients, when executing `dagger-engine-session`, also pass to it the address of the OCI image it was downloaded from. This is not included in the DSI spec. Is that a problem, and if so, how to resolve it?


Yeah.. it doesn't feel right to make this part of the spec. Isn't there a way to make that remote URL automatically defined? so that way the `engine-session` would know what image to use out of the box :thinking:

**Comment by sipsma** [comment]
> Yeah.. it doesn't feel right to make this part of the spec. Isn't there a way to make that remote URL automatically defined? so that way the engine-session would know what image to use out of the box 🤔

Yes there should be, I think we can map the tagged git version in the binary to an engine image tag (similar to what we did previously w/ moby/buildkit). I think we'll still want to resolve the engine image tag to a sha so that we handle any cases where the tag mutates (unexpected but possible), which introduces some complication, but I think that should be alright.

Or, maybe even better, all the above probably becomes much easier if we don't distribute the `engine-session` as part of the engine image anymore (which creates circular dependencies) and instead just switch the SDKs to obtain the binary from e.g. s3 (which makes a compelling case for merging `engine-session` into the CLI one way or another). I think this makes more sense now that we are going down the route of fully embracing "local helper binary" whereas previously I viewed it more as a temporary crutch.

**Comment by helderco** [comment]
> In Go it's straightforward but I've been told it's less straightforward in python+nodejs

Just a quick clarification on this. The default for GraphQL clients is to connect over http so this is always **available for free**. It's easy enough to add support for unix sockets it just may take a few lines of code of our own. 

As an example, the Python GraphQL client we're using has [these transports](https://gql.readthedocs.io/en/stable/transports/index.html) (basically http and websockets). The underlying http libraries have support for unix sockets, we just need to write the transport for the GraphQL client for it, but it's easy enough to do.

This work is planned in #3694, if we decide to have that choice (I'm actually in favor). I'm even planning on creating a new transport for [httpx](https://github.com/dagger/dagger/issues/3693) anyway so unix support will already be at hand.

**Comment by marcosnils** [comment]
> Yes there should be, I think we can map the tagged git version in the binary to an engine image tag (similar to what we did previously w/ moby/buildkit). I think we'll still want to resolve the engine image tag to a sha so that we handle any cases where the tag mutates (unexpected but possible), which introduces some complication, but I think that should be alright.
> 
> Or, maybe even better, all the above probably becomes much easier if we don't distribute the `engine-session` as part of the engine image anymore (which creates circular dependencies) and instead just switch the SDKs to obtain the binary from e.g. s3 (which makes a compelling case for merging `engine-session` into the CLI one way or another). I think this makes more sense now that we are going down the route of fully embracing "local helper binary" whereas previously I viewed it more as a temporary crutch.

cc @aluzzardi defining this will become important for the release of the CLI as well. Let's try to converge towards a decision before the end of the week if possible :pray:

**Comment by sipsma** [comment]
> defining this will become important for the release of the CLI as well. Let's try to converge towards a decision before the end of the week if possible 🙏

Just talked with @marcosnils and @aluzzardi about the plans here and the overlap with CLI

## Immediate term
To get the CLI ready to ship asap we'll do this:
1. Engine images will also be tagged by the git commit
2. The git commit will also be burned into the engine-session binary
3. The default behavior of the engine-session binary will be updated to provision the engine-image from that git commit tag
4. The CLI will just use `engine.Start` directly again rather than going through the Go SDK. It will just default to 
   * @marcosnils I don't think this was something explicitly mentioned in discussion before, just realized as I was typing this, let me know if you disagree though. I think it's overall simpler and makes sense in context of our new plans.

## Next Step (worked on in parallel ideally)
This is the end state we want, but whether we get it shippable in time for the release of the CLI is borderline due to the need to update each SDK; if we don't have it ready then it can be made in a follow up release:
1. The engine-session binary no longer exists inside the engine image.
   * This eliminates the weird circular dependency where a component of the server (engine-session) is brought locally and then used to provision the server.
2. Instead, the functionality of the engine-session binary will now be part of the CLI
   * TODO: determine precise details (e.g. do we just use `dagger listen` with a few extra knobs added or do we make the engine-session functionality a hidden subcommand, etc.)
3. When an SDK starts it will:
   * By default, dynamically provision a CLI binary matching its engine dependency
      * This is similar to how the dynamic provisioning works today, except now it will download the CLI binary from somewhere rather than obtain it from the engine image
   * Also have the optional ability to use an existing pre-installed CLI binary through the `bin://` engineconn
      * To start, `bin://` will be needed by our dagger-in-dagger tests
      * Users who don't like the "magical dynamic binary" feature can also use it
      * In the long term, once we have a better story around version compatibility we could also make this the default (and only use the dynamic provisioning as a fallback, with the option of disabling it entirely)
4. The work we did in the `Immediate term` section above will not go to waste though, the default engine image provisioning can continue to work exactly how it's described there, it just will be handled entirely in the CLI now and re-used by each SDK.

---

The overlap of the above with the proposal in this issue is mostly around the `Open questions & issues` section. There's still obviously a fair number of details to be filled in, just wanted to write out this high level sketch of the proposed direction first.

**Comment by sipsma** [comment]
Talked about a bit more this with @aluzzardi, synthesizing that with the plans on
1. replacing the engine-session binary with the CLI as ([described here](https://github.com/dagger/dagger/issues/3830#issuecomment-1317694885))
2. the various user requests for more customization of the runner ([here](https://github.com/dagger/dagger/issues/3939) and [here](https://github.com/dagger/dagger/pull/3943)).

I've tweaked a few things since Andrea and I talked, so blame me not him if this makes no sense :-)

### Important background facts:
1. We have now already diverged from the official `moby/buildkit` image (the fix to no longer build the shim dynamically). Users now have to use either our engine image or one based on it.
   * We expect this divergence to grow over time, including relatively soon to support networking configuration.

## Proposed Env Vars
These env vars will exist:
### `_DAGGER_SESSION_PORT` + `_DAGGER_SESSION_URL`
   * These point to an existing running session that can be connected to, as described above
   * There are no default values if these are not set
   * They are interpreted by each SDK (not CLI)
### `_DAGGER_CLI_BIN`
   * If the above env vars are **NOT** set, then the SDK will start the session itself from a binary pointed to by the env var.
   * Example values of this are:
      * `bin://<path to bin>` - use a local binary
      * `http://<url to binary>` - download a binary (e.g. from S3)
   * Defaults to our CLI binary in S3 if not set
   * This is interpreted by each SDK (not CLI)
### `_DAGGER_RUNNER_HOST`
   * Points to a runner to use for builds
   * Example values:
      * `docker-image://` - pull an image and start it with default options
      * `docker-container://` - use an existing docker container, expected to be already running
      * `unix://`, `tcp://`, - connect to an existing runner served at this address
      * `podman-container://`, `kube-pod://`, etc. (other existing buildkit connhelper types)
    * This has a default value *set in the CLI* of `docker-image://<engine image tagged with git commit>`
    * This is interpreted by the CLI (not SDK)

## Use Cases
All of these use cases will be supported on an experimental basis, hence the `_` preceding each env var. Users should expect that they might change in the future. They will be documented in internal docs (i.e. markdown in our repo) with similar stability warnings.

Once we feel comfortable that the architecture and approach has settled enough we can start more officially supporting this with actual docs, operator guides and finalized env vars.

### User who wants to use a custom image
1. Pull our engine image, apply any needed changes (e.g. include certs), push to your own registry
3. Execute dagger commands or SDKs with `_DAGGER_RUNNER_HOST=docker-image://<your custom image ref>`

They will need to ensure compatibility between the CLI and their custom image for now.

### User who wants to deploy their own runner
1. Set up the runner however you want
2. Execute dagger commands or SDKs with `_DAGGER_RUNNER_HOST=`, pointing to whatever runner you have set up (see valid values of `_DAGGER_RUNNER_HOST`

**Comment by shykes** [comment]
Thanks guys, the context & requirements are crytstal clear and I agree with them.

Comments on the proposed solution below.

> ## Proposed Env Vars
> These env vars will exist:
> 
> ### `_DAGGER_SESSION_URL`
> * This points to an existing running session that can be connected to, as described above in this issue and as implemented in [the previous PR for this](https://github.com/dagger/dagger/pull/3748#issuecomment-1317671329).
> * There is no default value if this is not set
> * It is interpreted by each SDK (not CLI)

Same comment as earlier in the thread. I believe the added complexity of handling a URL is not worth it. I would rather that we choose one of these options, and stick to it:

* Option 1: standardize on unix socket. `DAGGER_SESSION_SOCKET` (no leading `_` needed, as this would be permanent once set)
* Option 2: standardize on localhost tcp: `DAGGER_SESSION_PORT` (also no leading `_` for the same reason).

> ### `_DAGGER_CLI_BIN`
> * If the above env vars are **NOT** set, then the SDK will start the session itself from a binary pointed to by the env var.
> * Example values of this are:
>   
>   * `bin://<path to bin>` - use a local binary
>   * `http://<url to binary>` - download a binary (e.g. from S3)
> * Defaults to our CLI binary in S3 if not set
> * This is interpreted by each SDK (not CLI)

I think we should drop this entirely. It's too much complexity for clients, and not needed.

* `bin://` can be advantageously replaced with `$PATH`
* `http://` is not needed. As an end-user, either I want to rely on my client's auto-download feature as-is (no configuration needed), or I want to bypass it entirely (download myself + set `$PATH`).

> ### `_DAGGER_RUNNER_HOST`

Definitely keep a leading `_` in this one :)

> * Points to a runner to use for builds
> * Example values:
>   
>   * `docker-image://` - pull an image and start it with default options
>   * `docker-container://` - use an existing docker container, expected to be already running
>   * `unix://`, `tcp://`, - connect to an existing runner served at this address
>   * `podman-container://`, `kube-pod://`, etc. (other existing buildkit connhelper types)
>     * This has a default value _set in the CLI_ of `docker-image://<engine image tagged with git commit>`
>     * This is interpreted by the CLI (not SDK)

I think we should remove `docker-image://`, and anything else that doesn't point to a live endpoint. Probably safer to stick to the same set of supported values as buildkit, for now anyway.

Instead of `docker-image://`, just encode default behavior in the CLI *if this variable is not set*. The extra complexity of making the auto-provisioning logic configurable is not worth it. Anyone who cares enough to change the image, is likely to also want to control the provisioniong process anyway - no point in supporting this awkward in-between.

> ## Use Cases
> All of these use cases will be supported on an experimental basis, hence the `_` preceding each env var. Users should expect that they might change in the future. They will be documented in internal docs (i.e. markdown in our repo) with similar stability warnings.

Reading this now, I feel like a leading `_` is too easily dismissed. I think `_EXPERIMENTAL_` will be more explicit and clear, to minimize hard feelings down the road.

> Once we feel comfortable that the architecture and approach has settled enough we can start more officially supporting this with actual docs, operator guides and finalized env vars.
> 
> ### User who wants to use a custom image
> 1. Pull our engine image, apply any needed changes (e.g. include certs), push to your own registry
> 2. Execute dagger commands or SDKs with `_DAGGER_RUNNER_HOST=docker-image://<your custom image ref>`
> 
> They will need to ensure compatibility between the CLI and their custom image for now.

Per my comments below, this use case would change to be the same as the following one (but with `_EXPERIMENTAL_DAGGER_RUNNER_HOST`)

**Comment by aluzzardi** [comment]
> _DAGGER_SESSION_URL
>> Same comment as earlier in the thread. I believe the added complexity of handling a URL is not worth it. I would rather that we choose one of these options, and stick to it:

I think It's the opposite: underlying GraphQL libraries **do** require a URL. Not giving a URL pushes down the complexity of building one to the SDK.

- `_DAGGER_SESSION_URL`

```go
graphql.NewClient(os.Getenv("_DAGGER_SESSION_URL"))
```

- `DAGGER_SESSION_PORT`

```go
graphql.NewClient(fmt.Sprintf("http://localhost:%s", os.Getenv("_DAGGER_SESSION_URL")))
```

- `DAGGER_SESSION_SOCKET`

```go
httpClient := &http.Client{
	Transport: &http.Transport{
		DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
			return net.Dial("unix", os.Getenv("DAGGER_SESSION_SOCKET"))
		},
	},
}

graphql.NewClient("http://dummy/", httpClient)
```

AFAIK it' a bit more complex for Python (/cc @helderco) and Node (/cc @slumbering) as not all libraries support unix sockets.

**Comment by sipsma** [comment]
> _DAGGER_CLI_BIN
> I think we should drop this entirely. It's too much complexity for clients, and not needed.
> 
>  * bin:// can be advantageously replaced with $PATH
>  * http:// is not needed. As an end-user, either I want to rely on my client's auto-download feature as-is (no configuration needed), or I want to bypass it entirely (download myself + set $PATH).


I get what you're saying here but there's some constraints that make it tough to totally remove this at the moment:
1. We don't think we should search $PATH by default and then just fallback to auto-download. The problem is that version compatibility is too unstable at the moment to default to finding whatever binary exists locally and hope it works. 
   * The auto-download has the appealing feature that it results in the a compatible version always being used
   * `bin://` is still desirable for cases where you want to force the auto-download to not run and are willing to ensure version compat on your own
2. In general, I think there will probably be obscure use cases where users want to specify a specific binary to use without having to modify their `PATH` (which can have other undesirable side effects sometimes)

However I agree we don't have to expose `http://` at the moment, so I am good with dropping that.

@shykes What do you think about leaving `_EXPERIMENTAL_DAGGER_CLI_BIN` in such that:
1. If not set, autodownload from s3 kicks in (not configurable)
2. If set to `bin://<path>` then you can specify a path to a local binary (or if `<path>` is relative then `PATH` will be searched)

Then in the future once we have better version compatibility+stability we can consider switching it to what you described where PATH is the default and auto-download is the fallback (and deprecate `_EXPERIMENTAL_DAGGER_CLI_BIN`)

**Comment by sipsma** [comment]
> I think we should remove docker-image://, and anything else that doesn't point to a live endpoint. Probably safer to stick to the same set of supported values as buildkit, for now anyway.

I think I'm okay with this. The reason `docker-image://` is exposed is just for making it easier to test provisioning end to end before there is an officially released engine image. But I think the direction those provisioning tests are going will be that they run periodically rather than on every PR, so maybe it's actually fine to not allow overriding this.

One thing, the support for this value should have no impact on SDKs; the implementation lives entirely in the CLI and thus only has to be implemented once and by the dagger dev team (rather than per SDK and by whoever the SDK dev is). So if the need does arise for this to be exposed again I don't think we should rule it out but I agree all else being equal we should try to avoid that if possible.

**Comment by shykes** [comment]
> > _DAGGER_SESSION_URL
> > > Same comment as earlier in the thread. I believe the added complexity of handling a URL is not worth it. I would rather that we choose one of these options, and stick to it:
> 
> I think It's the opposite: underlying GraphQL libraries **do** require a URL. Not providing a URL pushes down the complexity of building one to the SDK.

I guess it depends on the schemes we need to support in the URL. So far every time we've support a URL format, a variety of schemes have found their way in.

So when I say "supporting a URL", I am thinking of:

* unix sockets. `unix://` ? Is this supported out of the box?
* ssh. `ssh://`?
* custom handlers for command execution etc `bin://`? `docker-container://`?

> As in, SDKs do not need to handle URLs -- there's no added complexity there, it's a passthrough. However, underlying HTTP libraries _do_ need a URL. If we don't provide one directly, then it's the SDK responsibility to reconstruct a working URL to pass downstream.
> 
> * `_DAGGER_SESSION_URL`
> 
> ```go
> graphql.NewClient(os.Getenv("_DAGGER_SESSION_URL"))
> ```

This is only true if all URL schemes we need the SDK to support are also pass-through (unclear).

>  * `DAGGER_SESSION_PORT`
> 
> ```go
> graphql.NewClient(fmt.Sprintf("http://localhost:%s", os.Getenv("DAGGER_SESSION_PORT")))
> ```
> 
>  * `DAGGER_SESSION_SOCKET`
> 
> ```go
> httpClient := &http.Client{
> 	Transport: &http.Transport{
> 		DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
> 			return net.Dial("unix", os.Getenv("DAGGER_SESSION_SOCKET"))
> 		},
> 	},
> }
> 
> graphql.NewClient("http://dummy/", httpClient)
> ```

Assuming 1) we want SDKs to support connecting by unix socket, and 2) graphql client doesn't support that, then: your "simple URL passthrough" is really not that simple, since it needs to do the same as this, plus parse the URL scheme to know when it is needed.

> AFAIK it' a bit more complex for Python (/cc @helderco) and Node (/cc @slumbering) as not all libraries support unix sockets.
> 
> Same applies to other tooling -- `dagger exec -- curl $DAGGER_SESSION_URL`, `dagger exec -- gql-lint $DAGGER_SESSION_URL`, etc etc
> 
> Not passing a URL would mean users are responsible of building their own wrapper to go from `non-URL -> URL -> tool`

What's not clear to me is what kinds of URLs we would require supporting. This lack of clarity is part of the problem in relying on a url in the spec. It leaves part of the spec implicit and easy to misinterpret.

Are you saying we should pass a URL *that only supports http over tcp*?

**Comment by shykes** [comment]
> Are you saying we should pass a URL that only supports http over tcp?

Assuming the answer to the above is "yes", then this seems like the best option to me:

> > ```go
> > graphql.NewClient(fmt.Sprintf("http://localhost:%s", os.Getenv("DAGGER_SESSION_PORT")))
> > ```

This option has several advantages compared to "URL + implicit convention that it should only be http"

1) not "complex" in any meaningful way
2) completely unambiguous, unlike URL where what schemas are expected to be handled is implicit leading to constant doubts and misunderstandings (exhibit A: this thread)
3) removes the accidental footgun of targeting a remote address (ie. it's never a good idea to target a non-localhost address)

**Comment by shykes** [comment]
> > _DAGGER_CLI_BIN
> there's some constraints that make it tough to totally remove this at the moment:
> 
> 1. We don't think we should search $PATH by default and then just fallback to auto-download. The problem is that version compatibility is too unstable at the moment to default to finding whatever binary exists locally and hope it works.

That's fair, I guess this becomes a bigger problem if the `dagger` CLI also acts as the session helper. My initial proposal relied on the fact that, most likely, you will never have `dagger-engine-session` in your PATH unless you're specifically looking to hijack. Different story with `dagger` obviously, so: I agree.

> What do you think about leaving `_EXPERIMENTAL_DAGGER_CLI_BIN` in such that:
> 
> 1. If not set, autodownload from s3 kicks in (not configurable)
> 2. If set to `bin://<path>` then you can specify a path to a local binary (or if `<path>` is relative then `PATH` will be searched)

Can we drop the `bin://` prefix? :) Eg. `_EXPERIMENTAL_DAGGER_CLI_BIN=<path>`. To me it feels less weird without the open-ended "url with weird schemes" contraption. You need a bin, here is its path, the end.

> Then in the future once we have better version compatibility+stability we can consider switching it to what you described where PATH is the default and auto-download is the fallback (and deprecate `_EXPERIMENTAL_DAGGER_CLI_BIN`)

With the modification above (no `bin://`) I could see this possibly stabilizing in the future, too. Either way I like the idea of keeping an experimental prefix until we're sure :)

**Comment by sipsma** [comment]
> Can we drop the bin:// prefix? :) Eg. _EXPERIMENTAL_DAGGER_CLI_BIN=<path>. To me it feels less weird without the open-ended "url with weird schemes" contraption. You need a bin, here is its path, the end.

Yeah, if we drop `http://` then it's weird if the only valid scheme is `bin://`, may as well just skip it.

**Comment by aluzzardi** [comment]
> I guess it depends on the schemes we need to support in the URL. So far every time we've support a URL format, a variety of schemes have found their way in.

Fair. The limitation was in dagger-in-dagger (and extensions) and the fact the only way we can communicate from within a container is through unix sockets (can't "mount" tcp ports).

So it was `http` outside and `unix` inside.

Because of the limitations of Node/Python (unix sockets) and the added complexity (this thread), it was simplified to:

From outside a container it's http and from inside a container, the "shim" takes care of unix<->http/tcp conversion so it's still http.

> Are you saying we should pass a URL that only supports http over tcp?

I think that would be the simplest approach, yes. At least from an SDK point of view. If it's not HTTP over TCP natively we make it so behind the scenes (shim or whatever)

> custom handlers for command execution etc bin://? docker-container://?

Yeah. That was because `session url`, `cli bin` and `runner host` were all mixed together into a single `DAGGER_HOST` -- the meaning and ownership of it really depended on the scheme. Splitting them up (this thread) removes the ambiguity, no more magic schemes.

> Assuming the answer to the above is "yes", then this seems like the best option to me:
> ```go
> graphql.NewClient(fmt.Sprintf("http://localhost:%s", os.Getenv("DAGGER_SESSION_PORT")))
> ```

I think I'd prefer to pass a full http URL, but that works for me too. There's a few downsides I can think of, but nothing major:

- I typo'd the sprintf, there's slightly more to it (forgot **path**) -- it should have been `fmt.Sprintf("http://localhost:%s/query", os.Getenv("DAGGER_SESSION_PORT"))`
- Pushes down a bit of complexity to the end client (e.g. `curl http://localhost:${DAGGER_SESSION_PORT}/query`)
- Hardcodes a few constraints
  - From within a container, the endpoint must be in the same netns (which is the case right now)
  - Hardcodes the path
  - Hardcodes non-TLS (can't switch from `http` to `https` if we wanted to)

> That's fair, I guess this becomes a bigger problem if the dagger CLI also acts as the session helper. My initial proposal relied on the fact that, most likely, you will never have dagger-engine-session in your PATH unless you're specifically looking to hijack. Different story with dagger obviously, so: I agree.

Right! Unsure yet if `dagger-engine-session` and `dagger` should be the same.

> Can we drop the bin:// prefix? 

👍

**Comment by sipsma** [comment]
It seems like the only remaining points here are around `URL`/`PORT`/`SOCK`, so I did a little more digging+thinking about localhost vs. unix sockets.

## TCP+Localhost

One big realization: **having the session listen on localhost means that anyone on the machine can connect to the session and read secrets from it**. 
* This is simpler than what I previously mentioned in terms of being able to sniff localhost traffic but also much worse in that it doesn't require any capabilities.
* E.g. anyone with access to localhost could obtain a secret by connecting to the session, setting the secret as an environment variable in a container and then just getting the stdout of running `env`

So if we commit to using localhost then I think we also need to commit to finding some way of safeguarding the session. This could be:
1. At very minimum - some sort of private token associated with session passed as a header. Though if this is plaintext then this will still be vulnerable to anyone that can sniff localhost traffic
2. Better but more complicated - mutual TLS where the session is tied somehow to the client identity
   * I don't know how certs will get setup. I'm sure we could figure out something but it seems like it could be fairly complicated (happy to be wrong).

So, if we go this route, I think we should keep our options open and allow use of `https` instead of `http` like @aluzzardi said. Unless of course we implement https support right away and never "officially" support http.

**Pros:**
1. Simplest to get going
2. Platform-agnostic

**Cons:**
1. Hardest to make secure

## Unix Sockets

Unix sockets improve the security posture by allowing us to use user/group permissions to safeguard a session. That overall feels much simpler than trying to accomplish the equivalent with setting up tls certificates.

For language support:
* Go - easy
* Python - easy [according to Helder here](https://github.com/dagger/dagger/issues/3830#issuecomment-1316003509)
* NodeJS - looks like our current graphql client enables overriding the `fetch` implementation used ([example in their tests here](https://github.com/prisma-labs/graphql-request/blob/c17f82138e5c0e7960fd1fb2ba9ef439323d57dc/tests/general.test.ts#L344)), so while not free it seems like it may not be *too* bad? cc @slumbering 

So overall impression is that it's not as easy localhost+tcp but pretty doable in our current languages.

Then there's the question of how we get unix sockets to work on other platforms. MacOS supports them natively. On Windows, I asked Chris Hunt and it seems like named pipes are usually considered the Windows-equivalent of unix sockets.
* He indicated that the graphql client he used in his powershell SDK work probably would need to be updated or replaced, but that overall it's not likely to be a huge burden: https://discord.com/channels/707636530424053791/1044713082020176032/1044715089749614602

This would mean that we need to have a separate implementation in engine-session when running on windows that creates a named pipe instead of unix socket. And each of our SDKs would need something similar to connect to that named pipe when running on windows.

**Pros:**
1. Simplest to make secure

**Cons:**
1. Hardest to get going
2. Not platform agnostic - need separate windows implementation that uses named pipes

## Overall

As much as I don't like needing a separate windows implementation that uses named pipes in each SDK, I currently think we should go with unix sockets + windows name pipes.

The difference from before is the realization that localhost exposes the session (and associated secrets) to anyone on the client machine, which IMO is too big to ignore. And the only "good" solution (mTLS certs) feels far too complicated (again, I'd be happy to be wrong though).

So, at this point I'd change my vote to be using `DAGGER_SESSION_SOCKET` and just bite the bullet of making that switch.

cc @aluzzardi @shykes @vito  let me know if you agree.

**Comment by shykes** [comment]
Yeah I agree that unprivileged access to the socket is too big to ignore. I also agree that the unix socket option, although not perfect, is a good tradeoff.

On http+localhost:

- I think https is only a solution if made mandatory (no http allowed). And that is probably too impractical so, IMO, a no-go
- I like the session token better. It could be a header, and even better: a standard http auth-basic header so that all clients support it OOTB. Yes it leaves the sniffing problem open, but: I worry less about that. In theory a non-root user could have network sniffing capability but that seems very unlikely to me. More likely is that if you can sniff traffic, it's because you're root, which means you don't need dagger's privileges in the first place.

So, I see 2 viable options each with a different tradeoff:

1. Unix socket
2. http/localhost with session token

**Comment by sipsma** [comment]
> I like the session token better. It could be a header, and even better: a standard http auth-basic header so that all clients support it OOTB. Yes it leaves the sniffing problem open, but: I worry less about that. In theory a non-root user could have network sniffing capability but that seems very unlikely to me. More likely is that if you can sniff traffic, it's because you're root, which means you don't need dagger's privileges in the first place.

The other thing I realized after writing my comment (edited it with it later) is that the session token approach doesn't solve another pretty significant vulnerability:  any local directory/file accessible to the session will become accessible over localhost. Anyone with access to localhost can connect to the session and import a localdir accessible to the session and then access file contents within.
   * E.g. if the `sipsma` user starts a session, the session will be running with my user's permissions. So then `baduser` who is on the same machine (but doesn't have access to `sipsma` owned files/dirs) can connect to localhost, load `/home/sipsma/.ssh` and then read keys from there.

Basically, use of `dagger` will significantly loosen any pre-existing file-permission-based security boundaries users may rely on.

I *think* you could address that by making the token a shared secret somehow, such that the server rejects any requests that don't have the correct token (as opposed to just namespacing requests based on it). But that opens more cans of worms in terms of securely sharing it in both the `dagger exec` and `dagger listen` cases, making sure it's always cryptographically random, not leaked into logs, etc. etc. In that sense, I think it's simpler to just rely on unix socket permissions.

**Comment by sipsma** [comment]
Just to double check, I did the quickest possible verification that unix sockets will usable achievable w/ nodejs, got something working here: https://github.com/sipsma/dagger/blob/22339d5da9d8af38fcceec63b6630bcbe13365ae/sdk/nodejs/api/client.gen.ts#L48

Needs cleanup and all that but I think that'll be doable.

For named pipes on windows, it does look like nodejs supports them as sockets in general: https://nodejs.org/api/net.html#identifying-paths-for-ipc-connections

[There also is/was support for making http requests over them in the `requests` library](https://github.com/request/request/issues/2296#issuecomment-527418999), but that library is apparently in maintenance mode now, so probably need to find another option there. So while windows support may require some extra effort it does seem doable too.

**Comment by sipsma** [comment]
One idea for a possible follow-up to consider if we go down the unix socket path: we *could* eventually still support an optional mode where `dagger listen` serves localhost rather than a unix socket, but in that localhost mode all requests are routed to run sandboxed in dagger-in-dagger rather than directly on the host.

This would not be the default behavior (`dagger listen` would use a unix sock by default) and is not something I propose we do immediately. But it is nice to keep something like `dagger listen --sandboxed` as an option on the table in case switching `dagger listen` to use unix sockets by default ends up being limiting.

**Comment by shykes** [comment]
> The other thing I realized after writing my comment (edited it with it later) is that the session token approach doesn't solve another pretty significant vulnerability: any local directory/file accessible to the session will become accessible over localhost. Anyone with access to localhost can connect to the session and import a localdir accessible to the session and then access file contents within.
> 
> * E.g. if the `sipsma` user starts a session, the session will be running with my user's permissions. So then `baduser` who is on the same machine (but doesn't have access to `sipsma` owned files/dirs) can connect to localhost, load `/home/sipsma/.ssh` and then read keys from there.
> 
> Basically, use of `dagger` will significantly loosen any pre-existing file-permission-based security boundaries users may rely on.
> 
> I _think_ you could address that by making the token a shared secret somehow, such that the server rejects any requests that don't have the correct token (as opposed to just namespacing requests based on it). But that opens more cans of worms in terms of securely sharing it in both the `dagger exec` and `dagger listen` cases, making sure users know to handle it very carefully from their code importing an SDK (never log it, unset it before passing to child processes if in env), ensuring it's always cryptographically random, etc. etc. In that sense, I think it's simpler to just rely on unix socket permissions.

I actually assumed that "session token" meant a shared secret for access control (no access if you don't have the token). I didn't even think about the namespacing idea...

And yes I agree, unix socket seems simpler.

**Comment by shykes** [comment]
Assuming we have consensus on `DAGGER_SESSION_SOCK`, that leaves us with the problem of `dagger listen`. The whole point of that feature is to offer a convenience for integrating graphql clients that 1) expect a "regular" remote graphql endpoint, and 2) can't easily be wrapped with `dagger exec / run`.

Basically, either we find a reasonably secure way for `dagger listen` to listen on a tcp port (ideally configurable by the user), or it becomes useless and we are better off cutting it.

My suggestion:

* We keep `dagger listen`
* `dagger listen` listens on a tcp address (localhost:0 by default, configurable by the user)
* A random path is generated each time, and printed to the user. Queries not prefixed by the random port are ignored.

Example:

```console
$ dagger listen -h 0.0.0.0:8080
GraphQL server listening on http://127.0.0.1:8080/f6a61bba
```

wdyt?

**Comment by sipsma** [comment]
@shykes I think `DAGGER_SESSION_PATH` is essentially operating as the secret session token we were talking about before, just now as part of the url path rather than as an http header.

My concerns with this are:
1. Clients should never log the url they send requests too as it is now secret. If they are using a generic graphql or http client and enable debug logging they might print this unintentionally
1. For the `dagger exec` case the secret is going to be transparently passed to applications through env, so users now need to know that, unset the env var before fork/execing, be careful to never print it, etc.

I feel like that's a lot of security footguns and a lot of documentation to push on the user.

> it becomes useless and we are better off cutting it.

I guess I agree but the more I think about it this is just the nature of what we're doing: we are serving an API that gives access to secrets, to local directories, to local env vars, etc. There's just no simple way to square that with an unsecured API running on localhost.

Honestly I think the idea of supporting a localhost mode that sandboxes requests via dagger-in-dagger could be the simplest route to making this secure:
1. Dagger-in-dagger inherently sandboxes the host dirs + env, so that's not a problem anymore
2. For the secret problem, we actually could perhaps rely on the "namespacing session token" idea mentioned earlier where the session token only needs to be randomly generated by the SDK and not shared through env
   * As long as this can be passed as an auth header (so generic http clients shouldn't log it) and doesn't need to be passed around env, I think that might be okay?

If we did want to go that route I think the missing parts would be:
1. Need a way of proxying host socket -> dagger-in-dagger exec (but we want that anyways as part of the whole services/networking thread)
2. Deciding on and implementing the session namespacing feature mentioned above

**Comment by shykes** [comment]
> @shykes I think `DAGGER_SESSION_PATH` is essentially operating as the secret session token we were talking about before, just now as part of the url path rather than as an http header.
> 
> My concerns with this are:
> 
> 1. Clients should never log the url they send requests too as it is now secret. If they are using a generic graphql or http client and enable debug logging they might print this unintentionally
> 2. For the `dagger exec` case the secret is going to be transparently passed to applications through env, so users now need to know that, unset the env var before fork/execing, be careful to never print it, etc.
> 
> I feel like that's a lot of security footguns and a lot of documentation to push on the user.

Are these really security footguns? We're talking about a path that is randomly generated for each session.

* For `dagger exec` it becomes useless as soon as the session is closed - typically within seconds. So I don't see any harm in printing it in the logs or leaking it anywhere.
* For `dagger listen`, in theory a user could decide to run it as a permanent production service. Then that path would become a problem. But the root problem there is that you should not use `dagger listen` for anything other than development and local integration with your own tools. Otherwise you will run into many problems, leaking this secret path being only one of them. This actually makes me think that maybe we should cut `dagger listen` out of caution.

> > it becomes useless and we are better off cutting it.
> 
> I guess I agree but the more I think about it this is just the nature of what we're doing: we are serving an API that gives access to secrets, to local directories, to local env vars, etc. There's just no simple way to square that with an unsecured API running on localhost.

Right. Ultimately `dagger listen` in its current form requires trusting that the user uses it responsibly.

> Honestly I think the idea of supporting a localhost mode that sandboxes requests via dagger-in-dagger could be the simplest route to making this secure:
> 
> 1. Dagger-in-dagger inherently sandboxes the host dirs + env, so that's not a problem anymore
> 2. For the secret problem, we actually could perhaps rely on the "namespacing session token" idea mentioned earlier where the session token only needs to be randomly generated by the SDK and not shared through env
>    
>    * As long as this can be passed as an auth header (so generic http clients shouldn't log it) and doesn't need to be passed around env, I think that might be okay?
> 
> If we did want to go that route I think the missing parts would be:
> 
> 1. Need a way of proxying host socket -> dagger-in-dagger exec (but we want that anyways as part of the whole services/networking thread)
> 2. Deciding on and implementing the session namespacing feature mentioned above

This is interesting and we may have a use for it at some point, but for `dagger listen` it defeats the purpose, which is basically to have a simple convenience to avoid having to type `dagger exec sh -c 'socat ...'`. At this point I'm inclined to say it's not worth it.

**Comment by shykes** [comment]
To summarize my preference after the last back and forth:

1. Cut `dagger listen`, not worth the complication
2. For DSI:
  2a. If it's in fact OK security-wise to pass a random path to avoid local port scanning, my first choice is `DAGGER_SESSION_PORT` + `DAGGER_SESSION_PATH`
  2b. If it's *not* OK security-wise, my second choice is `DAGGER_SESSION_SOCK`

**Comment by sipsma** [comment]
> Are these really security footguns? We're talking about a path that is randomly generated for each session.

I agree with your points here when it concerns "common cases" right now. 99.99% of the time right now there would be no issue.

But I'm also thinking about what happens when Dagger takes over the world and everyone is using it: the 0.01% "uncommon cases" (such as running long sessions or doing other unusual things) will start to statistically occur often enough and we will also become an attractive target for more sophisticated, specialized attacks that could maybe be made to work even in the common cases of short lived sessions. I'm basically just assuming the worst case scenarios will pop up eventually and seeing what we can do now to be resilient against them. 

> Cut dagger listen, not worth the complication

SGTM (cc @marcosnils)

>  2a. If it's in fact OK security-wise to pass a random path to avoid local port scanning, my first choice is DAGGER_SESSION_PORT + DAGGER_SESSION_PATH
> 2b. If it's not OK security-wise, my second choice is DAGGER_SESSION_SOCK

I'm still of the opinion that unix sockets will just be a simpler security model to work with. I hate that it'll require using named pipes w/ windows and thus platform-specific branching of our code, but I think it's worth it overall. There's just too many details around the port+session path to think about. And when it comes to this sort of thing I really don't like having lots of details to think about :-) With unix sockets, it's pretty much just `chmod 0600` and done.

I've started working on the change here to switch us to use unix sockets and will continue down this path for now just so I don't get blocked waiting for consensus. But if you feel strongly and/or others come in and also prefer `DAGGER_SESSION_PORT` + `DAGGER_SESSION_PATH` is preferable then I'm happy to be overruled and switch to the other approach (and accompanying documentation of the security considerations around treating that url carefully)

**Comment by sipsma** [comment]
# Summary of previous thread

| Env Var | Interpreted by | Description | Default Behavior if not set | Example Values
| --- | --- | --- | --- | --- |
| `DAGGER_SESSION_SOCKET`  | SDKs  | Points to a unix socket on which an existing session is being served | Fallthrough to check <br /> `_EXPERIMENTAL_DAGGER_CLI_BIN` | `/run/dagger/dagger-session-<randomid>.sock`
| `_EXPERIMENTAL_DAGGER_CLI_BIN`  | SDKs | Points to a CLI binary to use to start the engine-session | Pull CLI from S3 to local cache dir, use that  | `/usr/bin/dagger`
| `_EXPERIMENTAL_DAGGER_RUNNER_HOST`  | CLI | Points to a runner, schema indicates how to connect to it. Essentially our new `BUILDKIT_HOST` | Pull and run engine image in docker | `docker-container://my-dagger-engine`

* There are still some differing opinions about whether we should use `DAGGER_SESSION_SOCKET` vs. `DAGGER_SESSION_PORT` + `DAGGER_SESSION_PATH`. I'm going with SOCKET for now but this is still not 100% finalized.
* On Windows, we'll need to use named pipes rather than unix sockets, but that works with the above proposal still (just set `DAGGER_SESSION_SOCKET` to e.g. `\\.\pipe\PipeName`)

---

I did just realize that there's still one outstanding thread left, which is that `engine-session` functionality is assumed to be merged into the CLI as part of all of the above.

To re-iterate, the main goal is just to disentangle engine-session from needing to be pulled out of the engine image like happens today, which causes too many chicken-egg loops. To do that, we want to put it in S3. But since we are already pushing the CLI to S3 and the CLI is essentially the client-side of the engine "server", it makes sense to de-duplicate and just have the SDKs use it instead of trying to also manage a separate engine-session bin.
* It also creates a carrot to pre-installing the CLI and using that if you don't want the magic autodownload.

### My preferred option
Make `engine-session` a hidden subcommand of the CLI, especially since we're now planning on cutting `dagger listen` from the public interface to start.
* It can be hidden from the help output and also only invokable from a scary sounding name anyways like `dagger _internal_engine_session` so people seeing it in e.g. `ps` output know it's not meant to be used directly

### Fallback option
The other fallback is to just not combine these binaries yet. We can also publish `engine-session` to S3 separately for now.

**Comment by shykes** [comment]
> Make `engine-session` a hidden subcommand of the CLI, especially since we're now planning on cutting `dagger listen` from the public interface to start.

Makes sense, I think we could call it `dagger session`. Could you summarize the intended flags and behavior? Depending on that, we could even expose it to users as a slightly more low-level alternative to `dagger listen` perhaps?

**Comment by sipsma** [comment]
> Makes sense, I think we could call it dagger session. Could you summarize the intended flags and behavior? Depending on that, we could even expose it to users as a slightly more low-level alternative to dagger listen perhaps? 

Flags:
* `--workdir` - the workdir for the session (SDKs expose this as an option to `dagger.Connect`, it gets passed by the SDK to this binary)
   * The fact that this is a flag is just a historic artifact of copying code from cloak cli, it could be an env var too I suppose.

Env vars:
- `_EXPERIMENTAL_DAGGER_RUNNER_HOST` - as described above

Behavior:
1. Create and bind a unix socket at `$XDG_RUNTIME_DIR/dagger/dagger-session-<randomid>.sock`
   * set permissions to `0600` so only the same user can connect to it
2. Start the engine (including remote provision), serve the api at the unix sock
3. Write `$XDG_RUNTIME_DIR/dagger/dagger-session-<randomid>.sock` to stdout
   * same as before where we wrote the port number
4. Serve API until one of these things happens:
   * SIGINT/SIGTERM is received
   * stdin is closed (used to ensure this process dies when parent process dies without using Linux-specific features like pdeathsignal)
5. Logs will be written to stderr; the parent process is responsible for reading them and either forwarding them to where the SDK requested they go or just discarding them

---

That's basically just how `engine-session` works today except adjusted for the unix socket approach. It's optimized for machine consumption rather than user, but I guess it probably is technically usable by a human if they really want, so yeah we don't necessarily need to hide it.

**Comment by sipsma** [comment]
There was another round of discussion and updates yesterday:
1. Avoid unix socks+npipes.
   * These work on SDKs (https://github.com/dagger/dagger/pull/3983) but they don't work out of the box w/ generic graphql clients.
   * That means that for `dagger run` (and eventually `dagger listen`) we want to support tcp endpoints, with basic auth as a safeguard
   * But supporting both tcp and unix+npipes is a lot, so we are just going to standardize on the tcp+http basic auth approach for now across both the CLI commands and the SDK session interface
2. As mentioned above, use http basic auth with a token unique to the session to protect against anyone scanning localhost and connecting to the session
   * This should be good enough for our use case since even though we don't use https we are just running over localhost
   * The downsides are minor and/or obscure enough to be acceptable for now:
      1. Users have to be careful with the `DAGGER_SESSION_TOKEN` env var they get w/ `dagger run`. Just needs to be documented, not much different than any other system that allows secrets to be passed via env
      2. Anyone with capabilities to sniff localhost traffic (either granted to their process or via a setcap binary) can obtain the token
      3. SDKs will have to do a tiny bit extra work to securely obtain the token over the stdout of the engine session subprocess

---

# Latest Summary

## Env Vars
| Env Var | Interpreted by | Description | Default Behavior if not set | Example Values
| --- | --- | --- | --- | --- |
| `DAGGER_SESSION_URL`  | SDKs  | Points to an endpoint on which an existing session is being served | Fallthrough to check <br /> `_EXPERIMENTAL_DAGGER_CLI_BIN` | `http://localhost:34567`
| `DAGGER_SESSION_TOKEN`  | SDKs  | An HTTP basic auth username  | Fallthrough to obtain from CLI engine-session subprocess | `some-random-uuid`
| `_EXPERIMENTAL_DAGGER_CLI_BIN`  | SDKs | Points to a CLI binary to use to start the engine-session | Pull CLI from S3 to local cache dir, use that  | `/usr/bin/dagger`
| `_EXPERIMENTAL_DAGGER_RUNNER_HOST`  | CLI | Points to a runner, schema indicates how to connect to it. Essentially our new `BUILDKIT_HOST` | Pull and run engine image in docker | `docker-container://my-dagger-engine`

## CLI subcommands
1. `dagger session` - will initially be a hidden subcommand that just implements the same interface as engine-session today except:
   1. It will now also send the session token over stdout in addition to the port. The SDKs will use that as a basic auth header.

**Comment by sipsma** [comment]
@shykes @aluzzardi @vito @marcosnils posted the latest summary in my previous comment. Any other concerns left? If not, I'll break this down into subtasks (including as just discussed w/ Marcos updating the CLI to use the engine directly rather than the go sdk).

This all also ties heavily into OX, but I'm gonna make separate issues for more proposals there as it starts to branch too far out from this issue.

**Comment by shykes** [comment]
> There was another round of discussion and updates yesterday:

Thanks for the summary! Some thoughts below. I also updated the top comment to be up-to-date with my understanding.

> 1. Avoid unix socks+npipes.

👍

> 2. As mentioned above, use http basic auth with a token unique to the session to protect against anyone scanning localhost and connecting to the session

👍

>      2. Anyone with capabilities to sniff localhost traffic (either granted to their process or via a setcap binary) can obtain the token

Agreed, that is a security downside that we should acknowledge and document. I do believe that having an unprivileged local user that 1) may be untrusted, 2) has sniffing capability, and 3) is not root, is incredibly unlikely. A shop that cares about that risk would likely either make sure non-root doesn't have sniffing capability (seems like a no-brainer to me), *or* restrict local access to the machine. (Unprivileged, untrusted users that can sniff local traffic are bad news regardless of dagger).

>      3. SDKs will have to do a tiny bit extra work to securely obtain the token over the stdout of the engine session subprocess

I updated the proposal to specify how that should be encoded. I went for a json object because that seemed the simplest to implement and also extend in the future. But we can bikeshed.

> # Latest Summary

I updated the diagrams in the top comment, can you re-check that we're on the same page on the flow?

> `DAGGER_SESSION_URL`	SDKs	Points to an endpoint on which an existing session is being served	Fallthrough to check `_EXPERIMENTAL_DAGGER_CLI_BIN`	`http://localhost:34567`

I left `DAGGER_SESSION_PORT` in my proposal, and added an "open question" section to help us discuss and decide. I do think _PORT is better.

> `DAGGER_SESSION_TOKEN`	SDKs	An HTTP basic auth username	Fallthrough to obtain from CLI engine-session subprocess	`some-random-uuid`

No issue here. One question out of curiosity: will it cause confusion that the secret is passed as a username instead of password? Is that common practice for access tokens? Since the original purpose was to avoid "weirdness" I thought I'd ask.

> `_EXPERIMENTAL_DAGGER_CLI_BIN`	SDKs	Points to a CLI binary to use to start the engine-session	Pull CLI from S3 to local cache dir, use that	`/usr/bin/dagger`
> `_EXPERIMENTAL_DAGGER_RUNNER_HOST`	CLI	Points to a runner, schema indicates how to connect to it. Essentially our new `BUILDKIT_HOST`	Pull and run engine image in docker	`docker-container://my-dagger-engine`

I think those will be superseded by flags & configuration for `dagger run` soon enough, but that's why we prefixed them with `_EXPERIMENTAL`, so no objection :)

**Comment by sipsma** [comment]
> I updated the proposal to specify how that should be encoded. I went for a json object because that seemed the simplest to implement and also extend in the future. But we can bikeshed.

JSON SGTM

> One question out of curiosity: will it cause confusion that the secret is passed as a username instead of password? Is that common practice for access tokens?

I initially did this because it's what Stripe does: https://stripe.com/docs/api/authentication?lang=curl

And then some subsequent googling suggested that it's mostly arbitrary whether it's `(empty username, non-empty password)` or `(non-empty username, empty password)` but occasionally some systems (client or server-side) want you to never have an empty username, so username-only is a fine default.

I guess if we ever want this process to serve multiple sessions (not sure that we do, but who knows) it also makes some logical sense in that, with a lot of hand waving, you could then say that a session is closer to a user (i.e. scoped data) than a password.

> I left DAGGER_SESSION_PORT in my proposal, and added an "open question" section to help us discuss and decide. I do think _PORT is better. 

I'm fine with this. Someday we'll have to deal with some minor annoyance around forcing the engine session to listen on both ipv4 and ipv6 loopback addresses (https://gist.github.com/aojea/94f6f483173641647c731f582e52f0b0) since we can't just tell the client which one to connect to when it's only the port, but that's fine, obscure and probably easy to address.

Other than that it's mostly arbitrary for me since we can start w/ port and expand if needed in the future. I guess it's less one-way-door-y in that sense.

cc @marcosnils in case you have any objections to switch to DAGGER_SESSION_PORT

> Which CLI to download

Broad answer is that it's when we've done an engine upgrade and they have tested that their client works with it. For more specifics, I think we should split that off into a follow up issue (one of several that I'll be making once we're in agreement on the rest)

> Could dagger listen be folded into dagger session?

Yeah I think they will, I'm just waiting til I actually get to the implementation to confirm we can combine them since there's a small risk they have some subtle incompatibility in the interface (in which case `session` can just stay hidden indefinitely, no major harm done).

> I updated the diagrams in the top comment, can you re-check that we're on the same page on the flow?

Yep, my only extreme nit is that in DSI advanced we don't need to "export to environment variables"

**Comment by shykes** [comment]
> > One question out of curiosity: will it cause confusion that the secret is passed as a username instead of password? Is that common practice for access tokens?
> 
> I initially did this because it's what Stripe does: https://stripe.com/docs/api/authentication?lang=curl
> 
> And then some subsequent googling suggested that it's mostly arbitrary whether it's `(empty username, non-empty password)` or `(non-empty username, empty password)` but occasionally some systems (client or server-side) want you to never have an empty username, so username-only is a fine default.
> 
> I guess if we ever want this process to serve multiple sessions (not sure that we do, but who knows) it also makes some logical sense in that, with a lot of hand waving, you could then say that a session is closer to a user (i.e. scoped data) than a password.

👍 Makes sense, thanks for the details.

> > Which CLI to download
> 
> Broad answer is that it's when we've done an engine upgrade and they have tested that their client works with it. For more specifics, I think we should split that off into a follow up issue (one of several that I'll be making once we're in agreement on the rest)

👍

> > Could dagger listen be folded into dagger session?
> 
> Yeah I think they will, I'm just waiting til I actually get to the implementation to confirm we can combine them since there's a small risk they have some subtle incompatibility in the interface (in which case `session` can just stay hidden indefinitely, no major harm done).

👍

> > I updated the diagrams in the top comment, can you re-check that we're on the same page on the flow?
> 
> Yep, my only extreme nit is that in DSI advanced we don't need to "export to environment variables"

You’re absolutely right, even worse it’s an unnecessary pollution of the client’s environment. I can’t remember why it seemed like a good idea a few hours ago. I blame sleep deprivation 😅

**Comment by RealHarshThakur** [comment]
> _EXPERIMENTAL_DAGGER_RUNNER_HOST CLI 

@sipsma  Does this mean that when we do a `dagger.Connect(...)` , there would be a parameter to configure where the remote runner is?  

>  docker-container://my-dagger-engine

I'm trying to avoid using Docker. Is it possible to make it TCP-based similar to how BUILDKIT_HOST is?

**Comment by sipsma** [comment]
@RealHarshThakur

> Does this mean that when we do a dagger.Connect(...) , there would be a parameter to configure where the remote runner is?

For now we're thinking of an environment variable that is set when calling `dagger.Connect`, will that work for you? We are trying to avoid putting experimental, unstable parameters in our arguments to `dagger.Connect`, but if an env var creates hardship let us know.

> I'm trying to avoid using Docker. Is it possible to make it TCP-based similar to how BUILDKIT_HOST is? Or if it is the Dagger engine that's being exposed, HTTP-based?

You should be able to avoid Docker, however you will need to run the dagger engine somewhere and that engine is distributed as a container image. While we haven't tested it, I can't image why it fundamentally wouldn't work with non-docker container runtimes (podman/lima/etc.).

The caveat (for now anyways) is that if you don't want to use docker you will need to set up the container yourself somewhere ahead of time and then set `_EXPERIMENTAL_DAGGER_RUNNER_HOST` to point to it. We should be able to support `tcp://` addresses pretty easily for that though.

Does that sounds like it'd work for you?

**Comment by RealHarshThakur** [comment]
Env var works :)


>  can't image why it fundamentally wouldn't work with non-docker container runtimes (podman/lima/etc.).
I plan to run it with containerd (orchestrate it with K8s in future, maybe). IIRC, Podman/lima work with `docker://` but containerd doesn't. 

>  We should be able to support tcp:// addresses pretty easily for that though.
Yep, that sounds great.  

Would the image be a bundled Dagger engine+buildkit daemon? If so, would it be similar to load balancing and scaling Buildkit so cache is still shared?

**Comment by shykes** [comment]
> > `_EXPERIMENTAL_DAGGER_CLI_BIN`	SDKs	Points to a CLI binary to use to start the engine-session	Pull CLI from S3 to local cache dir, use that	`/usr/bin/dagger`
> > `_EXPERIMENTAL_DAGGER_RUNNER_HOST`	CLI	Points to a runner, schema indicates how to connect to it. Essentially our new `BUILDKIT_HOST`	Pull and run engine image in docker	`docker-container://my-dagger-engine`
> 
> 👍 I think those will be superseded by flags & configuration for `dagger run` soon enough, 

Just double-checking that we’re on the same page on this one @sipsma @RealHarshThakur @aluzzardi  . 

Given the general trend of making `dagger run` the primary mechanism for operator control, I think a logical conclusion is that configuration of remote engine will follow the same path. This would mean (in the future) deprecating `_EXPERIMENTAL_DAGGER_RUNNER_HOST` in favor of `dagger run` & associated config flags/files/env vars.

> but that's why we prefixed them with `_EXPERIMENTAL`, so no objection :)

Still relevant :)

**Comment by sipsma** [comment]
Sent out the first part of implementing all this: https://github.com/dagger/dagger/pull/4119

I also updated the description of this issue w/ the task list needed to implement it fully (cc @shykes). I considered breaking each task into a separate issue but I don't really see much benefit since they need to be done in quick succession of one another, not really left in the backlog. Happy to break into separate issues if others prefer though.

**Comment by mircubed** [comment]
note to self - we should update this help thread once this is closed -https://discord.com/channels/707636530424053791/1031785076029128784
